### ISSUE TITLE
Cross-Reference Centroids

### ISSUE THREAD

**Issue by CShorten** [semi-technologies/weaviate#2087]
**Tl;dr** 
Vectorize a class with the average vectors of a cross-referenced class.

**Motivation**
Averaging dense vectors from Deep Learning models preserves the semantics of the cluster. We see applications for this in decomposing large things such as an entire Wikipedia Article with vectors from each of the Paragraphs. We can also use the centroid vectors for personalizing search results based on User-Item relations, such as User -- liked --> Item.

**Proposal**

1. Configure Weaviate to vectorize a Class with referenceCentroids in the schema definition under "vectorizer" with custom centroid calculation arguments interfaced with "moduleConfig". 

2. Query with Centroid Vectors using NearObject.

**Example -- Wikipedia Decomposition on (Article, hasParagraph, Paragraph)**

Configure Vectorizer

<img width="485" alt="Paragraph-Schema" src="https://user-images.githubusercontent.com/25864937/183107294-f793ca78-382f-44b8-be8f-d206eb982389.png">

NearObject Query

<img width="676" alt="Paragraph-nearObject" src="https://user-images.githubusercontent.com/25864937/183107379-3ee4b16b-6f45-4b9b-a989-da15f53d6b4c.png">

**Example -- Product Personalization on (User, liked, Item)**

Configure Vectorizer

<img width="423" alt="User-Item-Schema" src="https://user-images.githubusercontent.com/25864937/183107445-45f65867-88f5-46ab-af75-e810cab9b8d4.png">

NearObject Query

<img width="651" alt="Item-nearObject" src="https://user-images.githubusercontent.com/25864937/183107486-2fd836c4-8235-442d-9e82-dcf84a38c5ce.png">

**Considerations**

1. We would like the centroid to be auto-updated when a user uploads a new cross-reference. _We could also introduce an "autoUpdate" property to turn this on/off in the moduleConfig_.

2. We may want to extend this in the future by adding a where filter to which referenced items are used in calculating the centroid. This concept is illustrated in the image below:

<img width="612" alt="Where-in-Centroid" src="https://user-images.githubusercontent.com/25864937/183108014-2892a27b-e8e2-40c6-a112-c9befac21212.png">

**Thank you!**

**Comment by bobvanluijt** [comment]
Tagging myself to follow discussions on this issue...

**Comment by bobvanluijt** [comment]
Seen in the wild: https://www.reddit.com/r/MachineLearning/comments/wjgj2h/comment/ijip016/?utm_source=share&utm_medium=web2x&context=3

**Comment by etiennedi** [comment]
Thanks @CShorten, great proposal and this makes a ton of sense. I would like to address the following points:

## Updates

One remark to this point:

> We would like the centroid to be auto-updated when a user uploads a new cross-reference. We could also introduce an "autoUpdate" property to turn this on/off in the moduleConfig.

I think there are two things that can be updated:

1. The class that has the `referenceCentroid` vectorizer set is updated. This needs to be updated when you change the (outgoing) references. This should be very easy to do. Since the object is already being updated (to manipulate the outgoing refs), updating the vector as well should be trivial. 
2. There is another situation: If the referenced objects are being updated, the `referenceCentroid` object may become stale. This is not something that we can catch automatically as references are only stored in one direction. Possible remedies might include: a) The user has to update the centroid class manually and the centroid will be recalculated with the current vectors at update time. b) There is some sort of a bigger "reindex" process. This would also help for other indexing situations (e.g. inverted index) that may benefit from reindexing. Given the size of this feature, I think it is something that can be discussed separately but just wanted to name it as a possible long-term remedy here.

## Calculating a centroid

1. We need to define how a centroid is actually calculated. I would assume that it's just the mean of the vectors, but it would be good to make that explicit
2. Do all distance metrics support centroids, and if so what does the calculation look like. For `cosine` and `l2-squared`, it's easy to calculate them, but for example for `dot product` I don't know if calculating a geometric centroid retains the meaning of both. This may need additional validation. If it's not possible for all metrics, we could also limit this feature to those metrics where it is possible.

## Naming

All other vectorizer modules follow the `<media>2vec-<something>` naming convention. Should this one be called `ref2vec-centroid` or something along those lines?

**Comment by CShorten** [comment]
Hey Etienne, thanks so much for this analysis!

**Updates**
1 - Fantastic, sounds perfect!
2 - Hmm, this is a very interesting idea if we have some recursion on this as well, i.e. Class1 defined by centroid with Cross-Reference to Class2 which is defined by centroid with Cross-Reference to Class3... hopefully not something that will deter the more low hanging fruit use-cases imagined like users clicking "like" on e-commerce products.

**Calculating a centroid**
1 - This is an excellent point. I think it would be best to keep it simple out of the gate with an average. For recommendation applications, the next steps would probably be a weighted average with time-stamps -- so the most recent "likes" weigh the most on the centroid embedding. For representing complex objects, I'm really not sure yet -- not getting positive results with a testbed of averaging scientific papers into a single vector with a sliding window. Overall great discussion point. Thinking a simple mean calculation will get this into a useful state, but certainly not solved for every application case. Hopefully we an design in such a way to extend with learnings.

2- I will try to find something more on this, haven't been able to develop a great sense on when cosine and l2 are better/worse.

**Naming**
Love that, ref2vec sounds great!

I hope this response was useful, thanks again!

**Comment by etiennedi** [comment]
FYI: This is on the list for Weaviate v1.16 🎉

**Comment by bobvanluijt** [comment]
> FYI: This is on the list for Weaviate v1.16 🎉

Great!

**Comment by colemanhindes** [comment]
any updates on this?

**Comment by etiennedi** [comment]
Yup, we [just released this feature](https://weaviate.io/blog/2022/11/Weaviate-release-1-16.html).
